engines/wazero/wazero.go

```diff
@@ -60,201 +71,283 @@ type (
+	m *Module
+}
 
-	m := &Module{
-		store:           store,
-		module:          module,
-		hostCallHandler: hostCallHandler,
+// Write implements io.Writer by invoking the Module.writer or discarding if nil.
+func (s *stdout) Write(p []byte) (int, error) {
+	w := s.m.wasiStdout
+	if w == nil {
+		return io.Discard.Write(p)
 	}
+	w(string(p))
+	return len(p), nil
+}
 
-	if _, err := wazero.InstantiateHostModule(store, wazero.WASISnapshotPreview1()); err != nil {
-		return nil, err
+// New compiles a `Module` from `code`.
+func (e *runtime) New(code []byte, hostCallHandler wapc.HostCallHandler) (mod wapc.Module, err error) {
+	r := wazero.NewRuntime()
+	m := &Module{runtime: r, wapcHostCallHandler: hostCallHandler}
+	// redirect Stdout to the logger
+	m.sysConfig = wazero.NewSysConfig().WithStdout(&stdout{m})
+	mod = m
+
+	if m.wasi, err = r.InstantiateModule(wazero.WASISnapshotPreview1()); err != nil {
+		mod.Close()
+		return
 	}
 
-	if _, err := wazero.InstantiateHostModule(store, &wazero.HostModuleConfig{
-		Name: "env",
-		Functions: map[string]interface{}{
-			"abort": m.env_abort,
-		},
-	}); err != nil {
-		return nil, err
+	if m.assemblyScript, err = instantiateAssemblyScript(r); err != nil {
+		mod.Close()
+		return
 	}
 
-	if _, err := wazero.InstantiateHostModule(store, &wazero.HostModuleConfig{
-		Name: "wapc",
-		Functions: map[string]interface{}{
-			"__guest_request":     m.wapc_guest_request,
-			"__guest_response":    m.wapc_guest_response,
-			"__guest_error":       m.wapc_guest_error,
-			"__host_call":         m.wapc_host_call,
-			"__host_response_len": m.wapc_host_response_len,
-			"__host_response":     m.wapc_host_response,
-			"__host_error_len":    m.wapc_host_error_len,
-			"__host_error":        m.wapc_host_error,
-			"__console_log":       m.wapc_console_log,
-		},
-	}); err != nil {
-		return nil, err
+	if m.wapc, err = instantiateWapcHost(r, m.wapcHostCallHandler, m); err != nil {
+		mod.Close()
+		return
 	}
 
-	return m, nil
+	if m.module, err = r.CompileModule(code); err != nil {
+		mod.Close()
+		return
+	}
+	return
 }
 
-// SetLogger sets the waPC logger for __console_log calls.
+// SetLogger implements the same method as documented on wapc.Module.
 func (m *Module) SetLogger(logger wapc.Logger) {
-	m.logger = logger
+	m.wapcHostConsoleLogger = logger
 }
 
-// SetWriter sets the writer for WASI fd_write calls to standard out.
+// SetWriter implements the same method as documented on wapc.Module.
 func (m *Module) SetWriter(writer wapc.Logger) {
-	m.writer = writer
+	m.wasiStdout = writer
 }
 
-// env_abort is the AssemblyScript abort handler
-func (m *Module) env_abort(ctx wasm.ModuleContext, messageOffset, fileOffset, line, col uint32) {
-	// TODO signal somewhere or call Close()? We don't need to indirect through WASI proc_raise I think
+// assemblyScript includes "Special imports" only used In AssemblyScript when a user didn't add `import "wasi"` to their
```

Comment:
I retained this even though other engines merely stub abort and don't do anything. I wonder if in 2022, we still have to stub this. Maybe this was made before WASI was available?